### ISSUE TITLE
Proposal: .NET Wrapper unnecessary heap allocations optimizations

### ISSUE THREAD

**Issue by MarsRover75** [zxing-cpp/zxing-cpp#838]
Hi.
A couple of thoughts about .NET wrapper:
[Read ](https://github.com/zxing-cpp/zxing-cpp/blob/535710a44a3f03f1fd51a908218a93640f496948/wrappers/dotnet/ZXingCpp/ZXingCpp.cs#L481) (and `From`) implementations are **always** returning `List<T>`, which is a reference object allocated on heap even if nothing is detected and it's empty.
```
public static List<Barcode> Read(ImageView iv, ReaderOptions? opts = null)
{
	var ptr = CheckError(ZXing_ReadBarcodes(iv._d, opts?._d ?? IntPtr.Zero));
	var size = ZXing_Barcodes_size(ptr);
	var res = new List<Barcode>(size);
	for (int i = 0; i < size; ++i)
		res.Add(new Barcode(ZXing_Barcodes_move(ptr, i)));
	ZXing_Barcodes_delete(ptr);
	return res;
}
```

when I'm reading QRs realtime from 50-120 FPS videostream this leads to a massive heap congestion.
Even this simple change makes a huge memory usage performance improvement in my case (since I'm using Server GC Mode and process videoframes in parallel):
```
public static Barcode[] ReadToArray(ImageView iv, ReaderOptions? opts = null)
{
	var ptr = CheckError(ZXing_ReadBarcodes(iv._d, opts?._d ?? IntPtr.Zero));
	var size = ZXing_Barcodes_size(ptr);
        if (size == 0) {
            // avoid unnecessary heap allocation, since it's just a reference to a static EmptyArray<T> instance.
            return Array.Empty<Barcode>();  // or just use collection expression []. 
        } 
	var res = new Barcode[size];
	for (int i = 0; i < size; ++i)
		res[i]= new Barcode(ZXing_Barcodes_move(ptr, i));
	ZXing_Barcodes_delete(ptr);
	return res;
}
```

But I went further.
I Extracted all public properties and methods of `Barcode` class to `IBarcode` interface, just to keep things uniform.
Then created my own `public struct BarcodeStruct : IBarcode, IDisposable`
Added some further optimizations, like additional properties that return `ReadOnlySpan<byte>` where original `Barcode` class properties return `byte[]`, same for properties that return `string`.

Then, I went for `public ref struct BarcodeReaderDiposableResults` that is returned by my `ReadDisposableResults(this ReaderOptions opts, ImageView iv)` static extension method for `ReaderOptions` class, the one, that actually calls `ZXing_ReadBarcodes` external C API.

It implements `Dispose()`, an indexer of type `BarcodeStruct` and also `GetEnumerator()` .
`GetEnumerator()` returns nested `public ref struct Enumerator`, which implements `public DisposableBarcodeStruct Current { get; }` and `public bool MoveNext()`. This way I can use `BarcodeReaderDiposableResults` in `foreach` loop statement.

Since `BarcodeReaderDiposableResults` is a `ref struct`, it's strongly scoped and can not be boxed, which means there are no implicit heap allocations, especially if I use my new `ReadOnlySpan<byte>` properties of `DisposableBarcodeStruct` to access the data I need.
Then I also turned `BarcodeStruct` into a ref struct just to be sure it's never boxed, pity currently ref structs that implement interfaces are a bit tricky, but It's going to [change soon](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/ref-struct#restrictions-for-ref-struct-types-that-implement-an-interface).

The code for this is a bit messy and needs cleanup, so I'm not posting it here, though it's quite simple, robust and zero-alloc.
I did some benchmarks on a real live videostream and it works great in terms of performance and memory usage (unless one forgets to call `Dispose()`, but I think this is a common concern when dealing with unmanageable resources).

([About disposable pattern with ref structs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/ref-struct#the-disposable-pattern))

**Comment by axxel** [comment]
Interesting stuff. I'm not a .NET developer myself and chose `List<>` as "the internet" suggested that this is the way to go here. What I find surprising is that you say `new Barcode[0]` and `new List<Barcode>(0)` behave fundamentally different in that the former actually does not do any heap memory allocation?

Please feel free to clean up your code and provide a PR that we can use as a basis to discuss any changes to the API. Luckily, I expressly declared the .NET wrapper "alpha" so far ;).

**Comment by MarsRover75** [comment]
> `new List<Barcode>(0)` behave fundamentally different in that the former actually does not do any heap memory allocation?
`new List<T>(0)` will **create** a new empty List object on a heap with every invokation, while Array.Empty<T>() will create it only once for every T and then just return a reference to it.
Arrays are immutable, Lists are not, that's the reason.

[Check it out in the source](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Array.cs,b23146dbd4923af8)

**Comment by axxel** [comment]
That makes sens, thanks for the link. In my opinion, returning a type that is immutable does generally make sens, too, from an API design perspective. `ReadOnlySpan` does also sound plausible, judged by the name. The ref Struct stuff, I have not looked into, yet.

**Comment by MarsRover75** [comment]
> The ref Struct stuff I have not looked into, yet.

`ref struct` type was basically "invented" for `Span<T>` historically, the point is to allow **safe** direct mapping to almost any memory region without use of pointers, that require unsafe context.
It brings huge performance benefits is used the right way, but has some limitations because of it's scoped lifetime, one can't just safely store and use a reference to a stack variable for example outside the stack frame.

**Comment by MarsRover75** [comment]
Just a little follow-up:
With .NET 7+ you can use compiler-generated strings marshallers for string return types instead of manually converting them (`Marshal.PtrToStringUTF8()`)
For example:
```
[LibraryImport(DllName, EntryPoint = "ZXing_Barcode_text")][return: MarshalAs(UnmanagedType.LPUTF8Str)] 
public static partial string ZXing_Barcode_text_as_string(IntPtr barcode);
```
notice that `LibraryImport` attribute is required is required in this case, not `DllImport`

**Comment by axxel** [comment]
Are you interested in providing a PR that changes the return type to `Barcode[]`?

If the `ReadOnlySpan` stuff for `text` and `bytes` is worth it, that might also be interesting. Here, it has to be taken into account that both members are passed as copies via the C-API already. And since they are only a couple bytes worth of data (in contrast to the image data that is passed into the library), I would question whether making the API more complex (like requiring the call to a `Dispose` function just to get rid of some string) is worth the prevention of one more copy.

**Comment by MarsRover75** [comment]
To put it simple:
Short-living wrapper objects like detection results and ImageView should be value types (struct, or ref struct) with the only real data member - `IntPtr` (it's a value type in .NET) to be passed to the unmanaged С counterpart, which in turn is just a ptr to some С++ class with it's data and vtable.

OR.
They might remain classes (red-types), but wrapper API should be able to reuse the instance of such a class as needed. A good example of this approach might be seen in Emgu.CV wrapper library, most of the managed reference object types like `Mat` can be instantiated once and then reused. 

Disposable pattern should be preferred over object finalizers when dealing with unmanaged resources anyway, it's just a general rule.

**Comment by axxel** [comment]
I replaced `List<Barcode>` with `Barcode[]` as you suggested. I noticed a memory leak in your above demo-code. The `ZXing_Barcodes_delete(ptr);` is missing in the `size == 0` case. Just in case this is still in your own implementation. Thanks for the input.

**Comment by MarsRover75** [comment]
> The `ZXing_Barcodes_delete(ptr);` is missing in the `size == 0` case. Just in case this is still in your own implementation. Thanks for the input.

Good point, but you're right, it's just a demo code.

My actual code uses only disposable ref structs and Results wrapper also implements IEnumerator interface, so I don't actually know the size in advance (why would I want to call `ZXing_Barcodes_size` unless it's  really needed right now?), all execution is deferred until results are enumerated.

**Comment by axxel** [comment]
Sounds like you take performance seriously. I respect that. In fact, most of the time I spent in total on this code base over the years was on performance improvements (and I take some [pride](https://github.com/axxel/zxing-bench/blob/main/dotnet/README.md) in that, too ;)).

That said, I question the value of potentially saving a few InterOp calls on the tail end of an image processing pipeline that executed millions of instructions. E.g. deferring the call to `ZXing_Barcodes_size` because it might never be called, makes no sens to me what so ever. If you never call that function, then you had better not called `ZXing_ReadBarcodes` in the first place ;).

But I'm curious: can you provide some benchmarking data to compare (the current) .NET wrapper and your improved version when scanning some real-world image data?

**Comment by MarsRover75** [comment]
> E.g. deferring the call to `ZXing_Barcodes_size` because it might never be called, makes no sens to me what so ever.

Haha, sure it's not that important with `size`, I was just trying to explain the deferred execution pattern in general.

> can you provide some benchmarking data to compare (the current) .NET wrapper and your improved version when scanning some real-world image data?

That would need the entire new project, since my current one heavily depends on my wrapper. It doesn't modify your wrapper codebase too much (except some access modifiers and `Dll` class declaration, which is `internal static partial class Dll` in my code), it just adds my classes/structs that replace the default implementation to it.

**Comment by MarsRover75** [comment]
My general point was - avoid re-allocations of data already allocated by C/C++ code, like QR payload **by wrapper**. 
If the managed client of the wrapper library wants to keep (which basically means copy) the data as it is - OK, go for it, there are many ways to do it. But do it explicitly in client code, the wrapper itself shouldn't allocate anything on the heap implicitly.

**Comment by axxel** [comment]
Since I fixed the most glaring issue (`List<Barcode>` vs. `Barcode[]`) but I have currently no plans to adopt your "deferred" API, I'm closing this now.
